pkg/autoscaler/k8sclient/k8sclient.go

```diff
@@ -140,17 +162,27 @@ type ClusterStatus struct {
 }
 
 func (k *k8sClient) GetClusterStatus() (clusterStatus *ClusterStatus, err error) {
-	opt := metav1.ListOptions{Watch: false}
-
-	nodes, err := k.clientset.CoreV1().Nodes().List(opt)
-	if err != nil || nodes == nil {
+	err = wait.PollImmediate(100*time.Millisecond, 5*time.Second, func() (bool, error) {
```

Comment:
I consciously copied it from addon-resizer. The reason is - what if it e.g. timesout?
1. With the current version - we will retry in the next attempt (in 10s by default)
2. With what you're suggesting, we would need to return error from client creation, and that would lead to restarting the whole component.

So I think this one is safer (and the overhead of checking a single string is negligible).

Reply:
That almost sounds reasonable, but I think that approach has a problem. If for some reason the component is unable to initialize the reflector for long time, it'll just keep trying and failing and retrying every 10s (running uselessly - but even worse, failing silently). OTOH if we fail the component right during startup (which I also feel is the ideal place to fail if the whole client hasn't initialized properly), the container will at least show up as 'failed' and we'll have a warning.

Reply:
If it's unable to initialize for long time, in majority (all?) of cases I can image it will also fail after restart.
Yes - restart is less silent, but noone really is looking into that, so I'm not sure it gives us anything signifcant.

Also note, that I can image that initialization may take some time. E.g. if apiserver has just crashed, or is being updated - it may take tens of seconds/minutes for it to start. We shouldn't be crashing this pod because of that immediately. 
So while i see your point, I think those are relatively weak ones.

Reply:
> Yes - restart is less silent, but noone really is looking into that, so I'm not sure it gives us anything signifcant.

No one may be looking into it currently, that doesn't mean no one would look into it in future :)
With your approach, the issue will be hidden (potentially forever) not even letting us get actionable signal. IMO having a fail-fast component is better than one which is silently failing forever. If the component is crashing, that will show up as 'container crash' which I believe is quite a visible signal for cluster admin (or automated health-checkers).
From a recent e.g - iiuc kubelet crashes within seconds in case it's client initialization, certificate-fetching, etc failed to happen during start instead of retrying again and again. I think this is also the case for few other control-plane components.

> E.g. if apiserver has just crashed, or is being updated - it may take tens of seconds/minutes for it to start.

I think there are 2 cases here:

- Apiserver has crashed due to a legit reason for multiple seconds - This has 2 further cases:
  - the autoscaler client didn't initialize and it crashed - This is justified for autoscaler component to crash as apiserver down is much much bigger problem (even kubelets will crash in such cases for e.g)
  - the autoscaler client initialized already - This is fine because it at least has some RV of the cache (it just won't be updated until apiserver comes up and won't do any scaling operation - which IMO is right behaviour)
- Apiserver is temporarily unavailable - we don't need to worry here as the component will eventually come up after some no. of restarts


Consider the e.g of kubelet, we fail it within seconds if it's

Reply:
I agree kubelet is a more serious component, but my point was about a more general pattern of "fail-fast".

> But that's not a disaster.

I don't fully agree. If we're unable to autoscale (which we could've by restarting the component), that might be a really bad experience as a customer whose unable to handle increased load from their clients.

I'm ok with changing it in a different PR, but I think we should at least add a TODO here. It seems to be a serious correctness issue if the reflector never succeeds to initialize.

Reply:
> (which we could've by restarting the component)

I can't imagine many cases where restarting would actually help. In huge majority of cases (all I've really seen somewhere?) - restarting wouldn't help.

I'm fine with a TODO, but that's more like "consider" todo, rather than "change" todo.